### ISSUE TITLE
Serving pipeline: replace LevelDB sorts with simple external sorter [phab:D631]

### ISSUE THREAD

**Issue by kythe** [google/kythe#892]
_Created by **schroederc** at 2015-12-05 04:13:34_: 

The intermediary stages for the partials edges, complete edges, and
decoration fragments only need to be sorted and then read sequentially.
Previously, this was done, for simplicity, with the existing LevelDB
machinery (carefully encoding each entry's key to get the correct
sorting), but it turns out that using a general key-value store for this
purpose is actually kind of slow.

This change introduces a new package, disksort, which was custom built
to take over LevelDB's job with less overhead.  Each element is now kept
in-memory (up to a specified number of elements) and then a set of data
is paged to disk as a sorted, delimited stream of bytestrings.  When all
elements have been added, the sorted shards are merged back together
incrementally and given to the user sequentially.

Quick Performance Numbers (write_tables --graphstore gs --out tbl):
  * 188,798 entries in GraphStore
    - Before: 7.20s
    - After:  5.85s

  * 13,970,489 entries in GraphStore
    - Before: 1562s
    - After:   548s

**Comment by kythe** [comment]
_**zarko** wrote at 2015-12-07 20:06:38_:

C++ test:

```
pre:
1715.40user 174.70system 12:38.07elapsed 249%CPU (0avgtext+0avgdata 2394736maxresident)k
712inputs+59225680outputs (0major+16382878minor)pagefaults 0swaps
post:
1937.21user 87.85system 7:06.28elapsed 475%CPU (0avgtext+0avgdata 1507640maxresident)k
3208inputs+33079360outputs (0major+6382655minor)pagefaults 0swaps
```

Looks roughly in line with what you have.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2015-12-07 21:37:05_:

nit: data [are] sharded (or is 'paged' more appropriate here?)

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2015-12-07 21:37:05_:

nit: number

**Comment by kythe** [comment]
_**zarko** wrote at 2015-12-07 21:37:05_:

Do you have any tests for `disksort`?

**Comment by kythe** [comment]
_**schroederc** wrote at 2015-12-07 23:17:42_:

Fix nits; fix error return in failure case; add disksort_test

**Comment by kythe** [comment]
_**schroederc** wrote at 2015-12-07 23:17:59_:

Sure

**Comment by kythe** [comment]
_**schroederc** wrote at 2015-12-07 23:18:54_:

Remove logging

**Comment by kythe** [comment]
_**zarko** accepted this revision at 2015-12-07 23:22:28_

**Comment by kythe** [comment]
_**schroederc** closed via commit 1035e95274e6481796115783694e1610cfc90b83 at 2015-12-08 00:04:40_
